source/ada/lsp-search.ads

```diff
@@ -0,0 +1,85 @@
+------------------------------------------------------------------------------
+--                         Language Server Protocol                         --
+--                                                                          --
+--                     Copyright (C) 2021, AdaCore                          --
+--                                                                          --
+-- This is free software;  you can redistribute it  and/or modify it  under --
+-- terms of the  GNU General Public License as published  by the Free Soft- --
+-- ware  Foundation;  either version 3,  or (at your option) any later ver- --
+-- sion.  This software is distributed in the hope  that it will be useful, --
+-- but WITHOUT ANY WARRANTY;  without even the implied warranty of MERCHAN- --
+-- TABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public --
+-- License for  more details.  You should have  received  a copy of the GNU --
+-- General  Public  License  distributed  with  this  software;   see  file --
+-- COPYING3.  If not, go to http://www.gnu.org/licenses for a complete copy --
+-- of the license.                                                          --
+------------------------------------------------------------------------------
+
+with VSS.Strings;  use VSS.Strings;
+with LSP.Messages; use LSP.Messages;
+
+package LSP.Search is
+
+   --------------------
+   -- Search_Pattern --
+   --------------------
+
+   type Search_Pattern is abstract tagged private;
+   type Search_Pattern_Access is access all Search_Pattern'Class;
+   --  A type used to describe the search context.
+   --  It can also be used to do the actual matching using the appropriate
+   --  algorithm, depending on the search kind.
+
+   procedure Free (Self : in out Search_Pattern) is null;
+   procedure Free (Self : in out Search_Pattern_Access);
+   --  Free the memory used by the matcher
+
+   function Build
+     (Pattern        : VSS.Strings.Virtual_String;
+      Case_Sensitive : Boolean := False;
+      Whole_Word     : Boolean := False;
+      Negate         : Boolean := False;
+      Kind           : Search_Kind := Full_Text)
+      return Search_Pattern_Access;
```

Comment:
We can return `Search_Pattern'Class` here and get rid of Free and any other memory management code.

Reply:
done